src/main/java/com/springclass/library/services/impl/LibraryServiceImpl.java

```diff
@@ -0,0 +1,196 @@
+        }
+
+        if (!isNumeric(id.toString())) {
+            throw new ServiceError(ErrorCode.INVALID_INPUT, "El ID del libro debe ser un valor numérico.");
+        }
+
+        if(!bookRepository.existsById(id)){
+            throw new ServiceError(ErrorCode.RESOURCE_NOT_FOUND,"Libro no encontrado");
+        }
+        bookRepository.deleteById(id);
+        return true;
+    }
+
+    @Override
+    public BookDTO createBook(CreateBookDTO input) throws ServiceError {
+
+        if (input.getTitle() == null || input.getAuthor() == null || input.getSubjects() == null) {
+            throw new ServiceError(ErrorCode.INVALID_INPUT, "Los parámetros no pueden ser nulos.");
+        }
+
+        Author author = getOrCreateAuthor(input.getAuthor());
+        List<Subject> subjects = getOrCreateSubjects(input.getSubjects());
+
+        Book book = new Book();
+        book.setAutor(author);
+        book.setTitle(input.getTitle());
+        book.setSubjectList(subjects);
+
+        bookRepository.save(book);
+
+        return convertBookDto(book);
+    }
+
+    private Author getOrCreateAuthor(String authorName){
+        Optional<Author> optionalAuthor = authorRepository.findByName(authorName);
+        return optionalAuthor.orElseGet(() -> {
+            Author newAuthor = new Author();
+            newAuthor.setName(authorName);
+            return authorRepository.save(newAuthor);
+        });
+    }
+
+    private List<Subject> getOrCreateSubjects(List<String> subjects) {
+        List<Subject> subjectList = new ArrayList<>();
+
+        for(String subject : subjects){
+            Optional<Subject> optionalSubject = subjectRepository.findByTitle(subject);
+            if (optionalSubject.isPresent()) {
+                subjectList.add(optionalSubject.get());
+            } else {
+                Subject newSubject = new Subject();
+                newSubject.setTitle(subject);
+                subjectList.add(subjectRepository.save(newSubject));
+            }
+        }
+        return subjectList;
+    }
+
+    @Override
+    public List<BookDTO> findBooks(BookSearch search) throws ServiceError {
+
+        if (search == null || search.getTitle() == null && search.getAuthor() == null && search.getSubject() == null) {
+            List<Book> allBooks = bookRepository.findAll();
+            return allBooks.stream()
+                    .map(this::convertBookDto)
+                    .toList();
+        }
+
+        List<Book> bookList = bookRepository.findAll();
+        List<Book> filterBookList = bookList.stream()
+                .filter(book -> isBookMatchingCriteria(book, search))
+                .toList();
+        return filterBookList.stream()
+                .map(this::convertBookDto).toList();
+    }
+
+    private boolean isBookMatchingCriteria(Book book, BookSearch search) {
+        boolean titleMatches = search.getTitle() == null || book.getTitle().toLowerCase().contains(search.getTitle().toLowerCase());
```

Comment:
Puedes jugar mucho con la clase `BookSearch` para crear filtros de búsqueda realmente potentes:
- Añadir atributos que determinen si el filtro debe cumplir todos los campos o solo alguno (AND/OR)
- Añadir paginación
- etc.
Si tienes tiempo, es un ejercicio muy interesante. Ten en cuenta que la paginación, búsqueda y filtrado es algo recurrente que se tiene que hacer en casi todas las APIs del mercado.

Reply:
el primer punto si puedo hacerlo, pero la paginacion tendria que hacerla en el metodo findbooks no? y tendria que cambiar los parametros de entradan no?

Reply:
Exacto, tendrías que cambiar no solo los argumentos de entrada, sino también los de salida y el método del controlador. No hace falta que lo hagas, aunque es recomendable que practiques estas técnicas. En cualquier caso, el próximo proyecto incluirá una búsqueda exhaustiva que requerirá que te des algunos cabezazos con estas cuestiones.